Global Precision/Recall/F1 Callback

jchen42703 opened the pull request:
## Description
* `PrecisionRecallF1ScoreMeter:` tracks TP, FP, and FN for each loader, and calculates precision (ppv), recall (tpr), and f1-score, based on those metrics.
* `PrecisionRecallF1ScoreCallback:` callback for `PrecisionRecallF1ScoreMeter`. Modeled after the AUCCallback & AccuracyCallback (multiple metrics).
* [Example logs for a 4 class multi-label classifier](https://drive.google.com/open?id=1dAfGca-IWlQibNbkenHdMkk6AF2niBF0)

## Related Issue
N/A

## Type of Change

<!-- Mark with an `x` all the checkboxes that apply (like `[x]`) -->

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
- [x] I have read the [CODE_OF_CONDUCT](../CODE_OF_CONDUCT.md) document.
- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) document.
- [x] I have checked the code-style using `make check-style`.
- [ ] I have written the docstring in Google format for all the method and classes that I used.
- [ ] I have checked the docs using `make check-docs`.

jchen42703 committed:
Created PrecisionRecallF1ScoreMeter

Keeps track of global true positives, false positives, and false negatives for each epoch and calculates precision, recall, and F1-score based on those metrics. Currently, for binary cases only (use multiple instances for multi-label).

jchen42703 committed:
Created PrecisionRecallF1ScoreCallback

Calculates the global precision (positive predictive value or ppv), recall (true positive rate or tpr), and F1-score per class for each loader. Currently, supports binary and multi-label cases.

jchen42703 committed:
Merge branch 'master' of https://github.com/jchen42703/catalyst into global-metrics

Updating with the latest changes to remove possibility of git diff bugs

jchen42703 committed:
Update __init__.py

For PrecisionRecallF1ScoreCallback

jchen42703 committed:
Added more documentation for the PrecisionRecallF1ScoreCallback's arguments

jchen42703 committed:
Fixing up codestyle for PEP 8

Passes `make codestyle`.

jchen42703 committed:
Fixed bug where the list of metrics was set as a float instead of appending the metric

This caused metrics to be averaged incorrectly because we were just calling np.mean(float) instead of np.mean(list of floats)

Scitator commented on catalyst/dl/meters/ppv_tpr_f1_meter.py:
```diff
@@ -0,0 +1,98 @@
+    Returns:
+        precision value (0-1)
+    """
+    return tp / (tp + fp + eps)
```
Could you please insert all used function in `PrecisionRecallF1ScoreMeter` on top of `PrecisionRecallF1ScoreMeter`? 
It would be nice to see everything before usage :)

Scitator replied:
btw, could you please also add some mini test cases, to check the implementation correctness?
(we have some fun stories already)

jchen42703 replied:
Just to clarify, do I make the test cases in dl/script and run it through an .sh file in catalyst/bin?

Scitator replied:
you can make some test like [these one](https://github.com/catalyst-team/catalyst/tree/master/catalyst/dl/tests)

Scitator commented on catalyst/dl/callbacks/metrics/ppv_tpr_f1.py:
```diff
@@ -0,0 +1,92 @@
+    def on_batch_end(self, state: RunnerState):
+        logits: torch.Tensor = state.output[self.output_key].detach().float()
+        targets: torch.Tensor = state.input[self.input_key].detach().float()
+        probabilities: torch.Tensor = torch.sigmoid(logits)
```
I think, we could parametrise the used function during callback initialization

jchen42703 replied:
I might be misinterpreting what you're saying, but are you asking to make a reusable function to do the ops above and initialize it as an attribute during callback initialization?

Scitator replied:
I mean, you can parametrize activation function like [here](https://github.com/catalyst-team/catalyst/blob/master/catalyst/dl/callbacks/metrics/iou.py#L17)

jchen42703 replied:
I think it would be more consistent, but I don't think `PrecisionRecallF1ScoreCallback` should inherit `MetricCallback`/`MultiMetricCallback` because `PrecisionRecallF1ScoreCallback` calculates the metrics for the entire loader instead of for each batch (maybe make a separate base global callback?). I'm curious on your thoughts?
(I'll add the option to specify the activation function but without the `super.init(...)` in the meantime)

Scitator replied:
I think we can create new abstraction like "DatasetMetricCallback", that need need to collect statistic during `on_batch_end` and also some additional calculations `on_loader_end`. We have the same case with AUCCallback, so... looks like we need something general :)

jchen42703 replied:
#450 <- Related PR

jchen42703 replied:
I'll try to get to it tomorrow!

smivv replied:
Do we need to add `.cpu()`? Otherwise it throws error:
`TypeError: can't convert CUDA tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.`

jchen42703 committed:
Moved f1score, precision, and recall to the top

jchen42703 committed:
Added the option to specify the activation function using catalyst.utils.get_activation_fn

jchen42703 committed:
Added more documentation for the .add(), .reset(), and .value() methods for PrecisionRecallF1ScoreMeter

Did so for increased clarity.

jchen42703 committed:
Created tests for precision, recall, f1score, and PrecisionRecallF1ScoreMeter

Still need to create tests for the callback, but this is a rough start.

jchen42703 committed:
Fixed up codestyle

Scitator commented on catalyst/dl/callbacks/metrics/ppv_tpr_f1.py:
```diff
@@ -0,0 +1,99 @@
+    def on_batch_end(self, state: RunnerState):
+        logits: torch.Tensor = state.output[self.output_key].detach().float()
+        targets: torch.Tensor = state.input[self.input_key].detach().float()
+        activation_fn = get_activation_fn(self.activation)
```
I think you can move it to init

jchen42703 committed:
Created MeterMetricsCallback

A callback that tracks metrics through meters and prints metrics for each class on `state.on_loader_end`. **Have not tested yet**

jchen42703 committed:
Added in missing imports, self.class_names, and spaces for codestyle

Should pass

jchen42703 committed:
Adapted PrecisionRecallF1ScoreCallback to inherit MeterMetricsCallback

Cleaner, reduces repeat code.

jchen42703 committed:
Adapted AUCCallback to inherit MeterMetricsCallback

Cleaner, reduces repeat code

jchen42703 committed:
Added MeterMetricsCallback for the .callback imports

jchen42703 committed:
Fixed codestyle

Removed unnecessary imports + overindented

jchen42703 committed:
Removed the section for when num_classes==1

Did so because catalyst does not currently support self.num_classes == 1 and len(probabilities.shape) == 1; tensors have no channels and you get a bunch of CUDA and memory pinning errors.

jchen42703 committed:
Made `class_names` and `num_classes` attributes for `MeterMetricsCallback`

Did so because doing so makes it clear to anyone implementing a child of `MeterMetricsCallback` that you need to specify `class_names` and `num_classes` for the callback to work properly.
Also added a check for num_classes == 1 (should be > 1).

jchen42703 committed:
Adapted the callbacks to work with the MeterMetricCallback's initialization changes (class_names and num_classes)

Did so for PrecisionRecallF1ScoreCallback and AUCCallback
